### PR TITLE
Implement RDM responder

### LINKED ISSUE #16: DMX RDM
Hi has anyone managed to implement RDM with this great library

### PR DESCRIPTION
**Pull request by someweisguy**
This adds the feature for the DMX library to respond to RDM requests using custom callbacks.

### CODE REVIEW

**Comment by someweisguy** [comment]
RDM response is working! The code still needs polish. Furthermore, the RDM response is useless for end-users; users are not able to configure settings on the RDM responder. This means that the RDM responder will always respond to messages the same way. But it supports the minimum required RDM responses, which is a pretty big milestone. 

The driver currently does not respond to invalid RDM requests, which would mean that RDM response is not yet fully supported in this library, but this should be trivial to implement.

**Comment by someweisguy** [comment]
I ran into some roadblocks with getting RDM response working so I decided to try a different solution. 

The new solution is a pretty big paradigm shift from how I was originally trying to implement RDM response. It took some work, but I got it working! It not only works, but the new paradigm is user-friendly enough for me to be happy. I am not quite finished with response, but I got enough of it working to see that it works very well. 

Unfortunately, due to the big conceptual changes made, I've had to break the RDM controller code. I am planning on re-implementing it before merging this PR. In any case, there is still lots of work that needs to be done before I can call RDM response finished.

**Comment by netmindz** [comment]
Sounds like good progress though

**Comment by someweisguy** [comment]
Discovery is working again. It's taking some time, but RDM is finally starting to come together. 

This PR is introducing a few breaking changes here and there so I am going to probably tag the next version v3.1 and officially take it out of beta. :)

**Comment by netmindz** [comment]
Let me know when the interface is stable enough for me to start trying to start trying to integrate

**Comment by someweisguy** [comment]
Okay, this is more or less feature complete for this PR. I'm working on documentation and polish now. But you are welcome to start messing around with code now - I won't be changing much from now on!

**Comment by netmindz** [comment]
Can you update the examples please?

**Comment by someweisguy** [comment]
> Can you update the examples please?

Thanks for your patience on this! I have been working primarily on documentation for the changes. I believe it is mostly complete. You can see the update documentation for RDM [here](https://github.com/someweisguy/esp_dmx/tree/feature/rdm-rx#reading-and-writing-rdm). Each function is also documented in their [header files](https://github.com/someweisguy/esp_dmx/tree/feature/rdm-rx/src/rdm).

Hopefully that is helpful, but I know that is not nearly as helpful as example code. I will be adding example code as soon as possible!

**Comment by netmindz** [comment]
Great to see you working on the docs, would be good to have the example added. You presumably have something already that you have been using for testing, so should just be a case of adding to the library?

**Comment by someweisguy** [comment]
[Here's](https://github.com/someweisguy/esp_dmx/blob/feature/rdm-rx/examples/idf_rdm_write/idf_rdm_write.c) the example code for RDM controller to get you started.

This example discovers RDM devices on the network, and gets the device info for each device. It prints the DMX start address and DMX footprint for each device. You should be able to swap out [these lines](https://github.com/someweisguy/esp_dmx/blob/feature/rdm-rx/examples/idf_rdm_write/idf_rdm_write.c#L49-L50) with a different request function found in [rdm/requests.h](https://github.com/someweisguy/esp_dmx/blob/feature/rdm-rx/src/rdm/requests.h) such as `rdm_get_software_version_label()`. The issue is that there are so many different RDM request functions that do very similar things with slightly different functions signatures that I'm thinking the best way to show example code is to add a function reference page to the readme. But maybe I'm overthinking things! :)

Edit: the Arduino example is [here](https://github.com/someweisguy/esp_dmx/blob/feature/rdm-rx/examples/RDMWrite/RDMWrite.ino)!

**Comment by netmindz** [comment]
Maybe I'm missing something, but the `examples/RDMWrite/RDMWrite.ino` and `examples/RDMDiscovery/RDMDIscovery.ino` both just seem to be trying to discover other RDM devices on the network, not actually be a responder using `rdm_register_callback` to return the device capabilities

**Comment by someweisguy** [comment]
> Maybe I'm missing something, but the `examples/RDMWrite/RDMWrite.ino` and `examples/RDMDiscovery/RDMDIscovery.ino` both just seem to be trying to discover other RDM devices on the network, not actually be a responder using `rdm_register_callback` to return the device capabilities

Nope you're not missing anything! This is my bad. I just added the `examples/RDMRead/RDMRead.ino` example. I was writing the comments for that example and I realized that it isn't actually a very user-friendly way to implement an RDM responder. I spent the last few days trying to see if I could make it a bit easier to use. I decided that it works, so that is good enough for this PR. I will be spending some time wracking my brain to see if I can update the API to make it easier to register callbacks in the future. 

Also, I haven't actually _tested_ this example yet. But the basic concept works. Let me know if you run into any issues!

**Comment by netmindz** [comment]
Yeah, I would agree at the moment it feels like you are exposing too much of the implementation in your interface.

While you might want to allow power users to do funky stuff, the default API would be much easier to use it if were more like this library, in general but definitely for the RDM stuff


https://github.com/chrisstaite/TeensyDmx/blob/master/examples/TeensyDMXRecv/TeensyDMXRecv.ino

**Comment by someweisguy** [comment]
Yeah, TeensyDmx looks excellent. It looks like the way TeensyDmx handles RDM controllers matches the way I've implemented RDM controller which is great to see. I'd just like to match that to the RDM responder code. I'll take another crack at it. 

There are also a few miscellaneous things that will need to be fixed in the RDM responder. Currently, for example, power-cycling the ESP32 would mean that the DMX start address is reset back to 1. It should hopefully be pretty straightforward to fix that.

**Comment by someweisguy** [comment]
Sorry for the radio silence. I spent some time redesigning a bit. The reason for this is because RDM supports many different parameters which have different message types so I wanted to avoid having to write a separate callback for each RDM parameter. Even so, there are a good number of edge cases that will still need to be considered in the response callbacks.

In any case, I made the response registry more user-friendly. The code is a bit bloated (and also throws a few warnings at compile time) so I want to spend some time trimming it down a bit and then I will work on documentation. In the meantime, I've provided the following functions:

```
bool rdm_register_device_info(dmx_port_t dmx_num, rdm_device_info_t *device_info);
bool rdm_register_software_version_label(dmx_port_t dmx_num, const char *software_version_label);
bool rdm_register_dmx_start_address(dmx_port_t dmx_num, uint16_t *dmx_start_address);
```

Users are able to register a pointer to a parameter and a callback will be registered which will handle an RDM controller getting or setting that parameter. The parameter is copied to the RDM driver by reference, not value. So the pointer to the parameter must be valid throughout the lifetime of the RDM driver. An easy way to do so is to declare your parameter variables static.

The following callback registry functions are also available:

```
bool rdm_register_disc_unique_branch(dmx_port_t dmx_num, void *context);
bool rdm_register_disc_mute(dmx_port_t dmx_num, void *context);
bool rdm_register_disc_un_mute(dmx_port_t dmx_num, void *context);
bool rdm_register_identify_device(dmx_port_t dmx_num);
```

These functions represent a few of those edge cases I mentioned above. I haven't fleshed out exactly what they should do yet. The three `_disc_` functions all work. `rdm_register_identify_device()` does not do anything currently. For now, don't touch these functions.

File: src/rdm/agent.c [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+  }
+  *uid = driver->rdm.uid;
+  taskEXIT_CRITICAL(spinlock);
+}
+
+
+void rdm_driver_set_uid(dmx_port_t dmx_num, rdm_uid_t uid) {
+  DMX_CHECK(dmx_num < DMX_NUM_MAX, , "dmx_num error");
+  DMX_CHECK(dmx_driver_is_installed(dmx_num), , "driver is not installed");
+  DMX_CHECK(uid_is_broadcast(&uid), , "uid error");
```
Comment (akares) [inline src/rdm/agent.c]: Hi! The check must be `!uid_is_broadcast` (not true instead of true).

File: src/rdm/mdb.c [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+  if (mdb && data) {
+    rdm_device_info_data_t *const pd = (void *)mdb->pd;
+    const rdm_device_info_t *param = data;
+    pd->major_rdm_version = 1;
+    pd->minor_rdm_version = 0;
+    pd->model_id = bswap16(param->model_id);
+    pd->product_category = bswap16(param->product_category);
+    pd->software_version_id = bswap32(param->software_version_id);
+    pd->footprint = bswap16(param->footprint);
+    pd->current_personality = param->current_personality;
```
Comment (akares) [inline src/rdm/mdb.c]: `pd->personality_count`  is missing here.

File: src/rdm/agent.c (hunk 1 above)
Comment (someweisguy) [reply src/rdm/agent.c]: Fixed! Thanks for catching this!

File: src/rdm/mdb.c (hunk 2 above)
Comment (someweisguy) [reply src/rdm/mdb.c]: Good catch here as well. I've added the missing line. 

That said, this function (and all the functions in `rdm/mdb.h`) will be removed soon and replaced with `rdm_encode()` in `rdm/utils.h`.

**Comment by netmindz** [comment]
I've been away for a while running a festival, but should hopefully have some time next week to test. 

Rather than diving straight into using this in WLED, in might send a PR with a fuller example and use that for my testing initially

**Comment by akares** [comment]
For testing I recommend this tool:
https://www.openlighting.org/rdm-tools/rdm-responder-tests/

It has more than 500 tests and nice coverage report.
Saves lots of time during development.

**Comment by someweisguy** [comment]
> For testing I recommend this tool: https://www.openlighting.org/rdm-tools/rdm-responder-tests/
> 
> It has more than 500 tests and nice coverage report. Saves lots of time during development.

This looks stellar. I have been using openlighting's RDM tools for testing, but I did not see that they have a bonafide RDM responder tester. I might have to shell out the cash to get a USB->DMX/RDM and try out the tester!

**Comment by someweisguy** [comment]
Okay, I'm feeling confident about the status of this PR now. There are still a few issues with how RDM packets are being parsed, but other than that, everything appears to be working. A few of the main changes are:

- RDM controller and responder functions must be included using either `#include "rdm/controller.h"` or `#include "rdm/responder.h"`.
- Sending RDM requests functions are now prefixed with `rdm_send_` instead of just `rdm_`. In other words, the function that was `rdm_get_dmx_start_address()` is now called `rdm_send_get_dmx_start_address()`. This is done to reduce confusion between getting another device's DMX start address, and your ESP32's DMX start address (which you do with `rdm_get_dmx_start_address()` now).
- Registering RDM responses are now done with functions prefixed with `rdm_register_`. If you'd like to register a callback to use a new RDM software version label, you can do so with `rdm_register_software_version_label()`. Up to 16 callbacks may be registered at a time. This hard limit may be adjusted in the `Kconfig`.
- The advanced RDM functions are found in `rdm/utils.h`. These would allow advanced users to do things most hobbyists don't really care for. For example, registering callbacks for different RDM sub-devices would be possible using these functions. This said, currently this PR cannot support sub-devices. I'm not planning on writing examples for the `rdm/utils.h` function for the time being, but I will eventually. If you're interested, I documented what these functions do in the header file.

I'm feeling like this is feature-complete enough to submit this PR, but there still a few kinks to work out that could be future PRs. I'll likely add a few more minor bug-fixes and cleanups to this PR in the next few days and then, if there is no feedback, merge it!  I've added examples for both an RDM controller and responder to play with the major functions. Let me know if there are any thoughts!

Edit to add: I will add Arduino examples soon!

**Comment by someweisguy** [comment]
Arduino examples are up! The examples are very messy. I will clean them up soon.

I also had the opportunity to test the RDM responder code using an ETC Colorsource 20. The Colorsource was able to detect and automatically patch the ESP32. The Colorsource was not able to send an `IDENTIFY_DEVICE` command to the ESP32, which was bizarre. After some testing, it appears that the Colorsource performed RDM discovery, sent requests for the ESP32's `DEVICE_INFO`, `SOFTWARE_VERSION_LABEL`, and `MANUFACTURER_LABEL`. I was able to quickly implement a response for the `MANUFACTURER_LABEL` so each of those requests were fine. The Colorsource also sent an RDM `QUEUED_MESSAGE` request, which this library does not currently support. I suspect that because the Colorsource did not receive the expected response to the `QUEUED_MESSAGE` request, it labelled the ESP32 as defective and refused to send the `IDENTIFY_DEVICE` request when prompted.

**Comment by akares** [comment]
> > For testing I recommend this tool: https://www.openlighting.org/rdm-tools/rdm-responder-tests/
> > It has more than 500 tests and nice coverage report. Saves lots of time during development.
> 
> This looks stellar. I have been using openlighting's RDM tools for testing, but I did not see that they have a bonafide RDM responder tester. I might have to shell out the cash to get a USB->DMX/RDM and try out the tester!

We are using ENTTEC DMX USB Pro and it works smoothly with the OLA software.
Are you located in US? I can ask my US colleague to order the USB->DMX adapter to your address.

**Comment by someweisguy** [comment]
> > > For testing I recommend this tool: https://www.openlighting.org/rdm-tools/rdm-responder-tests/
> > > It has more than 500 tests and nice coverage report. Saves lots of time during development.
> > 
> > 
> > This looks stellar. I have been using openlighting's RDM tools for testing, but I did not see that they have a bonafide RDM responder tester. I might have to shell out the cash to get a USB->DMX/RDM and try out the tester!
> 
> We are using ENTTEC DMX USB Pro and it works smoothly with the OLA software. Are you located in US? I can ask my US colleague to order the USB->DMX adapter to your address.

I am in the US. That is super generous of you! Thank you! You or your colleague can reach out to me at mitch@theweisbrods.com and I can send my mailing address.

**Comment by someweisguy** [comment]
Merged! There is still plenty of work to do on RDM. I'll be submitting more PRs to further polish up the RDM responder soon. For now I would say that RDM response is (minimally) supported.

**Pull request merged**